File: src/common/wasm/gather.ts
/**
 * WASM-accelerated extract and take_along_axis operations.
 */

import {
  extract_f64,
  extract_f32,
  extract_i64,
  extract_u64,
  extract_i32,
  extract_u32,
  extract_i16,
  extract_u16,
  extract_i8,
  extract_u8,
  take_axis0_2d_f64,
  take_axis0_2d_f32,
  take_axis0_2d_i64,
  take_axis0_2d_u64,
  take_axis0_2d_i32,
  take_axis0_2d_u32,
  take_axis0_2d_i16,
  take_axis0_2d_u16,
  take_axis0_2d_i8,
  take_axis0_2d_u8,
  where_f64,
  where_f32,
  where_i64,
  where_u64,
  where_i32,
  where_u32,
  where_i16,
  where_u16,
  where_i8,
  where_u8,
} from './bins/gather.wasm';
import {
  wasmMalloc,
  resetScratchAllocator,
  resolveInputPtr,
  scratchCopyIn,
  f16InputToScratchF32,
  f32OutputToF16Region,
} from './runtime';
import { ArrayStorage } from '../storage';
import { effectiveDType, type DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 64;

type ExtractFn = (condPtr: number, dataPtr: number, outPtr: number, N: number) => number;
type TakeFn = (
  dataPtr: number,
  indicesPtr: number,
  outPtr: number,
  rows: number,
  cols: number
) => void;

const extractKernels: Partial<Record<DType, ExtractFn>> = {
  float64: extract_f64,
  float32: extract_f32,
  int64: extract_i64,
  uint64: extract_u64,
  int32: extract_i32,
  uint32: extract_u32,
  int16: extract_i16,
  uint16: extract_u16,
  int8: extract_i8,
  uint8: extract_u8,
  float16: extract_f32,
};

const takeKernels: Partial<Record<DType, TakeFn>> = {
  float64: take_axis0_2d_f64,
  float32: take_axis0_2d_f32,
  int64: take_axis0_2d_i64,
  uint64: take_axis0_2d_u64,
  int32: take_axis0_2d_i32,
  uint32: take_axis0_2d_u32,
  int16: take_axis0_2d_i16,
  uint16: take_axis0_2d_u16,
  int8: take_axis0_2d_i8,
  uint8: take_axis0_2d_u8,
  float16: take_axis0_2d_f32,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
  int64: BigInt64Array,
  uint64: BigUint64Array,
  int32: Int32Array,
  uint32: Uint32Array,
  int16: Int16Array,
  uint16: Uint16Array,
  int8: Int8Array,
  uint8: Uint8Array,
  float16: Float32Array,
};

/**
 * WASM-accelerated extract (conditional gather).
 * condition must be flattened int32, data must be contiguous.
 * Returns ArrayStorage or null.
 *
 * Note: extract output size is unknown until kernel runs, so we use
 * wasmMalloc for worst-case, then trim. For the actual result we need
 * to know the count, so we read from the persistent region.
 */
export function wasmExtract(condition: ArrayStorage, storage: ArrayStorage): ArrayStorage | null {
  if (!condition.isCContiguous || !storage.isCContiguous) return null;

  const size = Math.min(condition.size, storage.size);
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = effectiveDType(storage.dtype);
  const kernel = extractKernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  // Condition must be a simple numeric type (not complex/bigint for the WASM kernel)
  const condDtype = condition.dtype;
  if (
    condDtype !== 'int32' &&
    condDtype !== 'float64' &&
    condDtype !== 'int8' &&
    condDtype !== 'uint8' &&
    condDtype !== 'int16' &&
    condDtype !== 'uint16' &&
    condDtype !== 'float32' &&
    condDtype !== 'uint32'
  )
    return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;

  // Convert condition to i32 for the WASM kernel
  const condOff = condition.offset;
  const condData = condition.data;
  const condI32 = new Int32Array(size);
  for (let i = 0; i < size; i++) {
    condI32[i] = condData[condOff + i] ? 1 : 0;
  }

  const outMaxBytes = size * bpe; // worst case: all selected

  // Use wasmMalloc for output (worst case)
  const outRegion = wasmMalloc(outMaxBytes);
  if (!outRegion) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  const condPtr = scratchCopyIn(condI32 as unknown as TypedArray);

  const isF16 = dtype === 'float16';
  const dataOff = storage.offset;

  let dataPtr: number;
  if (isF16) {
    dataPtr = f16InputToScratchF32(storage, size);
  } else {
    dataPtr = resolveInputPtr(
      storage.data,
      storage.isWasmBacked,
      storage.wasmPtr,
      dataOff,
      size,
      bpe
    );
  }

  const count = kernel(condPtr, dataPtr, outRegion.ptr, size);

  if (isF16) {
    const f16Region = f32OutputToF16Region(outRegion, count);
    outRegion.release();
    if (!f16Region) return null;
    return ArrayStorage.fromWasmRegion(
      [count],
      dtype,
      f16Region,
      count,
      Float16Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
    );
  }

  // The output region may be larger than needed. We create a view of just the count elements.
  // Since fromWasmRegion creates a view, the extra bytes are wasted but harmless.
  return ArrayStorage.fromWasmRegion(
    [count],
    dtype,
    outRegion,
    count,
    Ctor as unknown as new (buffer: ArrayBuffer, byteOffset: number, length: number) => TypedArray
  );
}

/**
 * WASM-accelerated take_along_axis for 2D arrays along axis 0.
 * Returns ArrayStorage or null.
 */
export function wasmTakeAlongAxis2D(
  storage: ArrayStorage,
  indices: ArrayStorage,
  axis: number
): ArrayStorage | null {
  if (axis !== 0) return null;
  if (storage.ndim !== 2 || indices.ndim !== 2) return null;
  if (!storage.isCContiguous || !indices.isCContiguous) return null;

  const [rows, cols] = storage.shape;
  const totalSize = rows! * cols!;
  if (totalSize < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = effectiveDType(storage.dtype);
  const kernel = takeKernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const outBytes = totalSize * bpe;
  const isF16 = dtype === 'float16';

  const outRegion = wasmMalloc(outBytes);
  if (!outRegion) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  let dataPtr: number;
  if (isF16) {
    dataPtr = f16InputToScratchF32(storage, totalSize);
  } else {
    dataPtr = resolveInputPtr(
      storage.data,
      storage.isWasmBacked,
      storage.wasmPtr,
      storage.offset,
      totalSize,
      bpe
    );
  }

  // Convert indices to i32
  const idxOff = indices.offset;
  const idxData = indices.data;
  const idxI32 = new Int32Array(totalSize);
  for (let i = 0; i < totalSize; i++) {
    idxI32[i] = Number(idxData[idxOff + i]);
  }
  const idxPtr = scratchCopyIn(idxI32 as unknown as TypedArray);

  kernel(dataPtr, idxPtr, outRegion.ptr, rows!, cols!);

  if (isF16) {
    const f16Region = f32OutputToF16Region(outRegion, totalSize);
    outRegion.release();
    if (!f16Region) return null;
    return ArrayStorage.fromWasmRegion(
      Array.from(indices.shape),
      dtype,
      f16Region,
      totalSize,
      Float16Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
    );
  }

  return ArrayStorage.fromWasmRegion(
    Array.from(indices.shape),
    dtype,
    outRegion,
    totalSize,
    Ctor as unknown as new (buffer: ArrayBuffer, byteOffset: number, length: number) => TypedArray
  );
}

// --- WASM where: out[i] = cond[i] ? x[i] : y[i] ---

type WhereFn = (condPtr: number, xPtr: number, yPtr: number, outPtr: number, N: number) => void;

const whereKernels: Partial<Record<DType, WhereFn>> = {
  float64: where_f64,
  float32: where_f32,
  int64: where_i64,
  uint64: where_u64,
  int32: where_i32,
  uint32: where_u32,
  int16: where_i16,
  uint16: where_u16,
  int8: where_i8,
  uint8: where_u8,
  float16: where_f32,
};

/**
 * WASM-accelerated element-wise where: out[i] = cond[i] ? x[i] : y[i].
 * All three arrays must be contiguous, same shape, non-complex, same dtype for x/y.
 * Condition is converted to i32 for the WASM kernel.
 * Returns ArrayStorage or null.
 */
export function wasmWhere(
  condition: ArrayStorage,
  x: ArrayStorage,
  y: ArrayStorage
): ArrayStorage | null {
  if (!condition.isCContiguous || !x.isCContiguous || !y.isCContiguous) return null;

  const size = condition.size;
  if (size !== x.size || size !== y.size) return null;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = effectiveDType(x.dtype);
  if (dtype !== y.dtype) return null;

  const kernel = whereKernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const dataBytes = size * bpe;
  const isF16 = dtype === 'float16';

  const outRegion = wasmMalloc(dataBytes);
  if (!outRegion) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  // Convert condition to i32
  const condOff = condition.offset;
  const condData = condition.data;
  const condI32 = new Int32Array(size);
  for (let i = 0; i < size; i++) {
    condI32[i] = condData[condOff + i] ? 1 : 0;
  }
  const condPtr = scratchCopyIn(condI32 as unknown as TypedArray);

  let xPtr: number;
  let yPtr: number;

  if (isF16) {
    xPtr = f16InputToScratchF32(x, size);
    yPtr = f16InputToScratchF32(y, size);
  } else {
    xPtr = resolveInputPtr(x.data, x.isWasmBacked, x.wasmPtr, x.offset, size, bpe);
    yPtr = resolveInputPtr(y.data, y.isWasmBacked, y.wasmPtr, y.offset, size, bpe);
  }

  kernel(condPtr, xPtr, yPtr, outRegion.ptr, size);

  if (isF16) {
    const f16Region = f32OutputToF16Region(outRegion, size);
    outRegion.release();
    if (!f16Region) return null;
    return ArrayStorage.fromWasmRegion(
      Array.from(x.shape),
      dtype,
      f16Region,
      size,
      Float16Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
    );
  }

  return ArrayStorage.fromWasmRegion(
    Array.from(x.shape),
    dtype,
    outRegion,
    size,
    Ctor as unknown as new (buffer: ArrayBuffer, byteOffset: number, length: number) => TypedArray
  );
}

File: src/common/wasm/zig/gather.zig
//! WASM gather/extract kernels for indexing operations.
//!
//! extract: copies elements where condition is nonzero (like np.extract).
//! take_along_axis_2d: gathers elements along axis 0 using index array (2D case).
//! Supported dtypes: f64, f32, i64, u64, i32, u32, i16, u16, i8, u8.

// --- extract: conditional gather ---
// Condition is always i32 (0 or nonzero). Data and output share a dtype.
// Returns the number of elements written to out.

/// Extract f64 elements where cond[i] != 0.
export fn extract_f64(cond: [*]const i32, data: [*]const f64, out: [*]f64, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract f32 elements where cond[i] != 0.
export fn extract_f32(cond: [*]const i32, data: [*]const f32, out: [*]f32, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract i64 elements where cond[i] != 0.
export fn extract_i64(cond: [*]const i32, data: [*]const i64, out: [*]i64, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract u64 elements where cond[i] != 0.
export fn extract_u64(cond: [*]const i32, data: [*]const u64, out: [*]u64, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract i32 elements where cond[i] != 0.
export fn extract_i32(cond: [*]const i32, data: [*]const i32, out: [*]i32, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract u32 elements where cond[i] != 0.
export fn extract_u32(cond: [*]const i32, data: [*]const u32, out: [*]u32, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract i16 elements where cond[i] != 0.
export fn extract_i16(cond: [*]const i32, data: [*]const i16, out: [*]i16, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract u16 elements where cond[i] != 0.
export fn extract_u16(cond: [*]const i32, data: [*]const u16, out: [*]u16, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract i8 elements where cond[i] != 0.
export fn extract_i8(cond: [*]const i32, data: [*]const i8, out: [*]i8, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

/// Extract u8 elements where cond[i] != 0.
export fn extract_u8(cond: [*]const i32, data: [*]const u8, out: [*]u8, N: u32) u32 {
    var idx: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) {
            out[idx] = data[i];
            idx += 1;
        }
    }
    return idx;
}

// --- take_along_axis for 2D along axis 0 ---
// For a [rows x cols] array and [rows x cols] index array (i32),
// out[r][c] = data[indices[r][c]][c].

/// Gather f64 elements along axis 0 for 2D array.
export fn take_axis0_2d_f64(data: [*]const f64, indices: [*]const i32, out: [*]f64, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather f32 elements along axis 0 for 2D array.
export fn take_axis0_2d_f32(data: [*]const f32, indices: [*]const i32, out: [*]f32, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather i64 elements along axis 0 for 2D array.
export fn take_axis0_2d_i64(data: [*]const i64, indices: [*]const i32, out: [*]i64, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather u64 elements along axis 0 for 2D array.
export fn take_axis0_2d_u64(data: [*]const u64, indices: [*]const i32, out: [*]u64, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather i32 elements along axis 0 for 2D array.
export fn take_axis0_2d_i32(data: [*]const i32, indices: [*]const i32, out: [*]i32, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather u32 elements along axis 0 for 2D array.
export fn take_axis0_2d_u32(data: [*]const u32, indices: [*]const i32, out: [*]u32, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather i16 elements along axis 0 for 2D array.
export fn take_axis0_2d_i16(data: [*]const i16, indices: [*]const i32, out: [*]i16, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather u16 elements along axis 0 for 2D array.
export fn take_axis0_2d_u16(data: [*]const u16, indices: [*]const i32, out: [*]u16, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather i8 elements along axis 0 for 2D array.
export fn take_axis0_2d_i8(data: [*]const i8, indices: [*]const i32, out: [*]i8, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

/// Gather u8 elements along axis 0 for 2D array.
export fn take_axis0_2d_u8(data: [*]const u8, indices: [*]const i32, out: [*]u8, rows: u32, cols: u32) void {
    var i: u32 = 0;
    const total = rows * cols;
    while (i < total) : (i += 1) {
        const col = i % cols;
        const row_idx: u32 = @intCast(indices[i]);
        out[i] = data[row_idx * cols + col];
    }
}

// --- count_nonzero for condition array ---

/// Count nonzero i32 elements. Used to pre-allocate extract output.
export fn count_nonzero_i32(cond: [*]const i32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (cond[i] != 0) count += 1;
    }
    return count;
}

// --- where: out[i] = cond[i] ? x[i] : y[i] ---

/// Element-wise where for f64: out[i] = cond[i] != 0 ? x[i] : y[i].
export fn where_f64(cond: [*]const i32, x: [*]const f64, y: [*]const f64, out: [*]f64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for f32.
export fn where_f32(cond: [*]const i32, x: [*]const f32, y: [*]const f32, out: [*]f32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for i64.
export fn where_i64(cond: [*]const i32, x: [*]const i64, y: [*]const i64, out: [*]i64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for u64.
export fn where_u64(cond: [*]const i32, x: [*]const u64, y: [*]const u64, out: [*]u64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for i32.
export fn where_i32(cond: [*]const i32, x: [*]const i32, y: [*]const i32, out: [*]i32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for u32.
export fn where_u32(cond: [*]const i32, x: [*]const u32, y: [*]const u32, out: [*]u32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for i16.
export fn where_i16(cond: [*]const i32, x: [*]const i16, y: [*]const i16, out: [*]i16, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for u16.
export fn where_u16(cond: [*]const i32, x: [*]const u16, y: [*]const u16, out: [*]u16, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for i8.
export fn where_i8(cond: [*]const i32, x: [*]const i8, y: [*]const i8, out: [*]i8, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

/// Element-wise where for u8.
export fn where_u8(cond: [*]const i32, x: [*]const u8, y: [*]const u8, out: [*]u8, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = if (cond[i] != 0) x[i] else y[i];
    }
}

// --- Tests ---

test "where_f64 basic" {
    const testing = @import("std").testing;
    const cond = [_]i32{ 1, 0, 1, 0 };
    const x = [_]f64{ 10.0, 20.0, 30.0, 40.0 };
    const y = [_]f64{ 1.0, 2.0, 3.0, 4.0 };
    var out: [4]f64 = undefined;
    where_f64(&cond, &x, &y, &out, 4);
    try testing.expectApproxEqAbs(out[0], 10.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 2.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 30.0, 1e-10);
    try testing.expectApproxEqAbs(out[3], 4.0, 1e-10);
}

test "where_u8 basic" {
    const testing = @import("std").testing;
    const cond = [_]i32{ 0, 1, 1, 0 };
    const x = [_]u8{ 10, 20, 30, 40 };
    const y = [_]u8{ 1, 2, 3, 4 };
    var out: [4]u8 = undefined;
    where_u8(&cond, &x, &y, &out, 4);
    try testing.expectEqual(out[0], 1);
    try testing.expectEqual(out[1], 20);
    try testing.expectEqual(out[2], 30);
    try testing.expectEqual(out[3], 4);
}

test "extract_f64 basic" {
    const testing = @import("std").testing;
    const cond = [_]i32{ 1, 0, 1, 0, 1 };
    const data = [_]f64{ 10.0, 20.0, 30.0, 40.0, 50.0 };
    var out: [3]f64 = undefined;
    const n = extract_f64(&cond, &data, &out, 5);
    try testing.expectEqual(n, 3);
    try testing.expectApproxEqAbs(out[0], 10.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 30.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 50.0, 1e-10);
}

test "extract_i8 basic" {
    const testing = @import("std").testing;
    const cond = [_]i32{ 0, 1, 0, 1 };
    const data = [_]i8{ 10, 20, 30, 40 };
    var out: [2]i8 = undefined;
    const n = extract_i8(&cond, &data, &out, 4);
    try testing.expectEqual(n, 2);
    try testing.expectEqual(out[0], 20);
    try testing.expectEqual(out[1], 40);
}

test "extract_u8 all true" {
    const testing = @import("std").testing;
    const cond = [_]i32{ 1, 1, 1 };
    const data = [_]u8{ 5, 10, 15 };
    var out: [3]u8 = undefined;
    const n = extract_u8(&cond, &data, &out, 3);
    try testing.expectEqual(n, 3);
    try testing.expectEqual(out[0], 5);
    try testing.expectEqual(out[2], 15);
}

test "take_axis0_2d_f64 basic" {
    const testing = @import("std").testing;
    // data: [[10, 20], [30, 40]]  indices: [[1, 0], [0, 1]]
    // out[0][0] = data[1][0] = 30, out[0][1] = data[0][1] = 20
    // out[1][0] = data[0][0] = 10, out[1][1] = data[1][1] = 40
    const data = [_]f64{ 10.0, 20.0, 30.0, 40.0 };
    const indices = [_]i32{ 1, 0, 0, 1 };
    var out: [4]f64 = undefined;
    take_axis0_2d_f64(&data, &indices, &out, 2, 2);
    try testing.expectApproxEqAbs(out[0], 30.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 20.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 10.0, 1e-10);
    try testing.expectApproxEqAbs(out[3], 40.0, 1e-10);
}

test "take_axis0_2d_i8 basic" {
    const testing = @import("std").testing;
    const data = [_]i8{ 1, 2, 3, 4, 5, 6 }; // 3x2
    const indices = [_]i32{ 2, 0, 1, 2, 0, 1 }; // 3x2
    var out: [6]i8 = undefined;
    take_axis0_2d_i8(&data, &indices, &out, 3, 2);
    try testing.expectEqual(out[0], 5); // data[2][0]
    try testing.expectEqual(out[1], 2); // data[0][1]
    try testing.expectEqual(out[2], 3); // data[1][0]
    try testing.expectEqual(out[3], 6); // data[2][1]
}

test "count_nonzero_i32 basic" {
    const testing = @import("std").testing;
    const cond = [_]i32{ 0, 1, 0, 1, 1, 0 };
    try testing.expectEqual(count_nonzero_i32(&cond, 6), 3);
}
